xds/src/main/java/io/grpc/xds/sds/FileBasedPluginCredential.java

```diff
@@ -0,0 +1,124 @@
+
+  @VisibleForTesting final String headerKey;
+  @VisibleForTesting final String headerPrefix;
+  @VisibleForTesting final DataSource secretData;
+
+  FileBasedPluginCredential(MetadataCredentialsFromPlugin metadataCredentialsFromPlugin) {
+    checkNotNull(metadataCredentialsFromPlugin, "metadataCredentialsFromPlugin");
+    checkArgument(
+            PLUGIN_NAME.equals(metadataCredentialsFromPlugin.getName()),
+            "plugin name should be %s", PLUGIN_NAME);
+    checkArgument(metadataCredentialsFromPlugin.hasConfig(),
+            "typed_config not supported");
+
+    Struct configStruct = metadataCredentialsFromPlugin.getConfig();
+
+    if (configStruct.containsFields(HEADER_KEY)) {
+      Value value = configStruct.getFieldsOrThrow(HEADER_KEY);
+      headerKey = value.getStringValue();
+    } else {
+      headerKey = DEFAULT_HEADER_KEY;
+    }
+    if (configStruct.containsFields(HEADER_PREFIX)) {
+      Value value = configStruct.getFieldsOrThrow(HEADER_PREFIX);
+      headerPrefix = value.getStringValue();
+    } else {
+      headerPrefix = "";
+    }
+    Value value = configStruct.getFieldsOrThrow(SECRET_DATA);
+    checkState(value.hasStructValue(), "expected struct value for %s", SECRET_DATA);
+    secretData = buildDataSourceFromConfigStruct(value.getStructValue());
+  }
+
+  private static DataSource buildDataSourceFromConfigStruct(Struct secretValueStruct) {
+    checkNotNull(secretValueStruct, "secretValueStruct");
+    if (secretValueStruct.containsFields(FILENAME)) {
+      Value value = secretValueStruct.getFieldsOrThrow(FILENAME);
+      return DataSource.newBuilder().setFilename(value.getStringValue()).build();
+    } else {
+      throw new UnsupportedOperationException("only secret_data of filename type supported");
+    }
+  }
+
+  @Override
+  public void applyRequestMetadata(
+      @Nullable RequestInfo requestInfo, Executor appExecutor, final MetadataApplier applier) {
+    appExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              Metadata headers = new Metadata();
+              String headerValue =
+                  headerPrefix
+                      + Files.asCharSource(new File(secretData.getFilename()), Charsets.UTF_8)
+                          .read();
+
+              if (headerKey.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+                Metadata.Key<byte[]> metadataHeaderKey =
+                    Metadata.Key.of(headerKey, Metadata.BINARY_BYTE_MARSHALLER);
+                headers.put(metadataHeaderKey, headerValue.getBytes(Charsets.UTF_8));
+              } else {
+                Metadata.Key<String> metadataHeaderKey =
+                    Metadata.Key.of(headerKey, Metadata.ASCII_STRING_MARSHALLER);
+                headers.put(metadataHeaderKey, headerValue);
+              }
+              applier.apply(headers);
+            } catch (IOException e) {
+              applier.fail(Status.fromThrowable(e));
```

Comment:
Status.fromThrowable(e) will be guaranteed to return UNKNOWN status, because it normally expects Status(Runtime)?Exception. 
```suggestion
              applier.fail(Status.UNKNOWN.withCause(e));
```

Reply:
I will leave this unchanged because `fromThrowable` does the same thing. `fromThrowable` could be enhanced (in the future) to also handle `IOException`s

Reply:
i left comment by knowing it does the same thing. i suggested this because doing this way delivers different message/expectation to the readers of this code.